### PR TITLE
OTLP encoder

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
Code + tests for OTLP encoding. Tests using OkHttp to send converted spans to Jaeger.

### CODE REVIEW

File: .github/workflows/create_release.yml [hunk 1]
```diff
@@ -1,13 +1,12 @@
-# yamllint --format github .github/workflows/create_release.yml
```
Comment (codefromthecrypt) [inline .github/workflows/create_release.yml:1]: can you pull the workflow updates, and the version bumps in docker and pom.xml into a PR, then rebase this? I would like to see only stuff about the encoder side and review that. cheers!

**Comment by marcingrzejszczak** [comment]
Sure I'll do that, but the thing is that starting from Monday I'm out for 10 weeks. I don't think the version bumps and the workflow updates would make it impossible to review this?

**Review by codefromthecrypt — COMMENTED** [review]
just to set expectations, I'm not going to review for merge something I can't merge. this does things that I warned about for a while (unrelated things in the same PR, let's please keep PRs coherent and easy to merge).

"OTLP encoder" should have a clear description of what it does, and not at the same time delete collector and other things like changing docker images etc.

totally understand if you are busy readying for a 10 week time off, we can pick up when you get back if you are not in a position to have only encoder changes in this PR.

Whatever we click approve on should be mergable, basically, let's make that the case even if it means waiting until you are not as busy.

**Comment by marcingrzejszczak** [comment]
OK, I've done everything that was requested of me, I've applied all the review suggestions. I think that there's no point in delaying the review of this any further.

File: encoder-otel-brave/pom.xml [hunk 2]
```diff
@@ -0,0 +1,160 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (codefromthecrypt) [inline encoder-otel-brave/pom.xml:1]: plase use spdx headers everywhere. the license plugin should catch this...

e.g. https://github.com/openzipkin/zipkin-gcp/blob/master/sender-pubsub/src/main/java/zipkin2/reporter/pubsub/PubSubSender.java#L1-L4

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/JaegerAllInOne.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * the License.
+ */
+package zipkin2.reporter.otel.brave;
+
+import java.util.Collections;
+import java.util.Set;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.HttpWaitStrategy;
+
+class JaegerAllInOne extends GenericContainer<JaegerAllInOne> {
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/JaegerAllInOne.java:21]: I assume there's no opentelemetry project that allows read-back of spans and that's why we are using jaeger. if so, comment that?

File: encoder-otel-brave/pom.xml [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    <dependency>
+      <groupId>${brave.groupId}</groupId>
+      <artifactId>brave</artifactId>
+      <version>${brave.version}</version>
+      <!-- Don't pin Brave -->
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.mock-server</groupId>
```
Comment (codefromthecrypt) [inline encoder-otel-brave/pom.xml:73]: are we using this? if not remove it

**Review by codefromthecrypt — COMMENTED** [review]
made a few notes as I have a flight to catch but made some time.

please try to adjust your tone with phrasing like you see no reason why X should do Y etc. this is volunteer work and I've been doing my best to instruct you clearly how to minimize. reviews are more likely to happen when folks are pleasant with each other.

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+  @Test
+  void shouldSendBraveSpansToJaegerOtlpEndpoint() throws Exception {
+    testingScenario = new JaegerTestingScenario(
+        jaegerAllInOne.getHttpOtlpPort());
+
+    List<String> traceIds = testingScenario.exportedTraceIds();
+
+    Awaitility.await().untilAsserted(() -> {
+      then(traceIds).hasSize(EXPECTED_TRACE_SIZE);
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:69]: we don't use BDD stuff in zipkin repos, let's follow style. e.g. `assertThat(foo).isBar();`

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    });
+  }
+
+  /**
+   * Sender: Brave OKHttp with OTLP proto over HTTP ; Receiver: Jaeger OTLP
+   */
+  class JaegerTestingScenario {
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:96]: there's no style like this in any of our other senders.  we don't need to introduce nouns like  JaegerTestingScenario. please remove this type

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+    public void close() {
+      braveTraceIdGenerator.close();
+    }
+  }
+
+  /**
+   * Actual testing logic that uses Brave to generate spans and send them to the backend.
+   */
+  static class BraveTraceIdGenerator {
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:120]: inline this into the test please. As there is no logic that behaves differently with different trace IDs, please use a static trace ID

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+      for (int i = 0; i < EXPECTED_TRACE_SIZE; i++) {
+        Span span = braveTracer.nextSpan().name("foo " + i)
+            .tag("foo tag", "foo value")
+            .kind(Kind.CONSUMER)
+            .error(new RuntimeException("BOOOOOM!"))
+            .remoteServiceName("remote service")
+            .start();
+        try (SpanInScope scope = braveTracer.withSpanInScope(span)) {
+          String traceId = span.context().traceIdString();
+          log.info("Trace Id <" + traceId + ">");
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:155]: remove this as there's only one test and one trace ID as a result of the desciption

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+      }
+
+      @Override
+      public SpanHandler apply(Integer port) {
+        okHttpSender = OkHttpSender.newBuilder()
+            .encoding(Encoding.PROTO3)
+            .endpoint("http://localhost:" + port + "/v1/traces")
+            .build();
+        OtelEncoder otelEncoder = new OtelEncoder(Tags.ERROR);
+        reporter = AsyncReporter.builder(okHttpSender).build(otelEncoder);
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:207]: why is this using AsyncReporter when it is AsyncSpanHandler that is used for brave?

**Review by codefromthecrypt — COMMENTED** [review]
thanks for the start, but there is a lot of cleanup to do here. Please focus on MutableSpan, AsyncSpanHandler and remove mentions of .V1 and make sure there is no use of zipkin2.Span or related types in this PR. I made a bunch of comments which will get this both smaller and into such a state.

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+  @Test
+  void shouldSendBraveSpansToJaegerOtlpEndpoint() throws Exception {
+    testingScenario = new JaegerTestingScenario(
+        jaegerAllInOne.getHttpOtlpPort());
+
+    List<String> traceIds = testingScenario.exportedTraceIds();
+
+    Awaitility.await().untilAsserted(() -> {
+      then(traceIds).hasSize(EXPECTED_TRACE_SIZE);
+      thenAllTraceIdsPresentInBackend(testingScenario.queryUrl(), traceIds);
```
Comment (codefromthecrypt) [inline encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:70]: one note I didn't mention was that this test isn't sufficient as an integration test as it doesn't verify the data was readable except its trace ID. I think I earlier said you could use an opentelemetry utility to deserialize a trace read back to verify mappings worked as we expected. So, in the test read-back the json such that the reader of the test will see if there is a miss. This is also why I suggest using a constant trace as it makes the verification idempotent, just subject to readback delay.

Meanwhile, I'll poke around to see if otel has a first party utility in the collector to read-back, or if everyone are using jaeger for this (which is possible)

**Comment by codefromthecrypt** [comment]
After noticing the image not changing from jaeger, I looked as it seemed odd otel would depend on jaeger to test their collector. As I suspected, there's another approach they use. Let's please use that instead as requested earlier

https://github.com/open-telemetry/opentelemetry-java/blob/main/integration-tests/otlp/src/main/java/io/opentelemetry/integrationtest/OtlpExporterIntegrationTest.java#L837

cc @anuraaga @reta in case i disappear until the following weekend and you feel like looking. Otherwise, next time there's a significant change here and I have time to spend an hour or two reviewing I will.

**Comment by marcingrzejszczak** [comment]
> please try to adjust your tone with phrasing like you see no reason why X should do Y etc. this is volunteer work and I've been doing my best to instruct you clearly how to minimize. reviews are more likely to happen when folks are pleasant with each other.

Thank you Adrian for this comment. I didn't want to offend you and I really appreciate the time you've decided to spend on reviewing this. Also thank you for the reminder about being pleasant to each other. I didn't mean to be unpleasant, I'm sorry. I'll try to improve my communication so that you don't feel uncomfortable. I'll look at the way you communicated with the community in the past for pointers on how to improve.

I though that now that you're the full-time OpenTelemetry Champion and looking back at your previous project engagements I assumed that having Brave work with OTel will be aligning with your work goals. Most likely I misinterpreted this situation for which I'm sorry.

Also the fact that I left Jaeger in this PR even though you asked me to remove it it's because I think it should stay there. That's a really good integration test IMO (of course I'll apply some of the other suggestions from this PR because they seem to be highly reasonable). Since this is a contrib repo and apparently I'm an admin here I understand that we can meet half ways with our expectations as far as the produced code is concerned.

As far as minimization is concerned I think that you're absolutely right and thank you for that remark. However extremes are never a solution. I'm pretty sure this PR (and actually the previous ones too) were minimal enough but of course we can have differences in opinions.

**Comment by codefromthecrypt** [comment]
@marcingrzejszczak I'm trying to understand what you are expecting as an outcome. It is a norm, to preserve the time of maintainers to do one feature at a time. Brave being a separate feature from collector etc is not an extreme, and it helps others less familiar with what's going on tell what is going on. Also, the whole point of this repo is to align with otel, so it doesn't seem an extreme to follow their practice and use otel not jaeger to test it, something I mentioned many times.

If you feel like this is your repo, because you have admin, I think you are missing the point. You don't have a long term history in maintaining things in this org and really can't say what is acceptable or not, or what is extreme or not. Of yours you can ignore the collaborators in an awkward way and make up your own rules, but I see no good coming out of that and I suspect you don't either.

I think basically you have a decision to make which is either to contribute empathetically and lean into feedback, decide to take this somewhere else, or let someone else implement this. Personally, I could have implemented this far quicker and easier than having to make feedback that costs time, and is received grudgingly or ignored. It would also be compatible in provable ways, not just because we can look up a trace ID from jaeger. The same result would be achievable from any contributor, and still can be possible from you.

Finally, from the very beginning I said very explicitly this is volunteer work on my side, so I'm not sure where you got the impression that this is a work task for me. Even if this was related to my job, no one can assign me work. In general, it takes a lot of time and patience from me to handle both your code and how you receive feedback.

I really do hope you can turn this around, but also don't blame you if you feel you're better off taking it somewhere else. If that is an otel org, I think you'll find they also have standards such as one feature as a time and you'll have to gain approval from an existing maintainer. Maybe the fastest way to code how you want with the least feedback is to work in your own repo, which is still possible should you feel that's the way out. Anyway, let us know how you want to proceed, as others can do this work in a straightforward fashion if you don't want to do it anymore.

**Comment by marcingrzejszczak** [comment]
I was planning to work on this during my leave but due to your tone and your attitude I've decided not to. What I can tell you Adrian is 

"please try to adjust your tone. this is volunteer work and I've been doing my best to instruct you clearly what I think makes best sense for the community. PRs are more likely to happen when folks are pleasant with each other."

@shakuzen and @jonatan-ivanov will continue working on this.

**Comment by codefromthecrypt** [comment]
@marcingrzejszczak ack your sentiments and understood.

@shakuzen or @jonatan-ivanov so if you are interested, my suggestion is to put the brave stuff into zipkin-otel-brave (as semantic mappings to brave-http, brave-rpc etc will follow I suspect). inside that module, make an encoder package. I would just make a new PR which focuses on MutableSpan to bytes as this did, but if you feel like modifying this you can.

My suggestion is to work backwards from the integration test and make the least possible code
* pre-canned mutableSpan -> existing sender + new encoder -> otlp -> readback span sent
* pre-canned mutableSpan -> existing sender + zipkin encoder -> zipkin endpoint on collector with zipkin enabled -> readback span sent

This way, we'll conserve time and get good coverage. On this PR I didn't get us to a point where we are comparing multiple spans, but eventually we would have to in order to test translation concerns a single client span cannot.

Hope these tips help, and please do affirmatively ack if either of you are indeed taking over as we don't want to block willing others if you can't.

**Comment by making** [comment]
I'd love to see Zipkin's OTEL support happen. May I volunteer to continue Marcin's work?

**Comment by codefromthecrypt** [comment]
@making looks like you have a go!

**Comment by making** [comment]
> After noticing the image not changing from jaeger, I looked as it seemed odd otel would depend on jaeger to test their exporter. As I suspected, there's another approach they use. Let's please use that instead as requested earlier (mtls is not required here unless otel's image requires it)
> 
> https://github.com/open-telemetry/opentelemetry-java/blob/main/integration-tests/otlp/src/main/java/io/opentelemetry/integrationtest/OtlpExporterIntegrationTest.java#L837
> 
> cc @anuraaga @reta in case i disappear until the following weekend and you feel like looking. Otherwise, next time there's a significant change here and I have time to spend an hour or two reviewing I will.

@codefromthecrypt 


It looks like `sender -> otel collector -> mock otlp receiver`. Would there be any problem with simply changing it to `sender -> mock otlp receiver` without going through the otel collector? This would eliminate the dependency on testcontainer. It seems easy to implement a simple mock otlp receiver (both http/gRPC) with Armeria.

**Comment by codefromthecrypt** [comment]
> It seems easy to implement a simple mock otlp receiver (both http/gRPC) with Armeria.

To me a mock with armeria would be more like a unit test `OtelEncoderTest` than an integration test `ITOtelEncoder`. You eventually need both as you may be verifying incompatible data in the test. Through the IT/testcontainers at least the shape of the data should be 100pct verified by the time you read it back, and there's possibly a processor which can do validation.

So, if you want to start with just a mock test, make an issue to complete the integration test. Make sense?

**Comment by making** [comment]
@codefromthecrypt Thanks. 
Maybe I misused the word "mock". What I meant was a server that accepts a byte array in a protocol buffer over HTTP and deserializes it into an `ExportTraceServiceRequest`, which is what I need to read back even via the otel collector. It's almost the same as https://github.com/open-telemetry/opentelemetry-java/blob/main/integration-tests/otlp/src/main/java/io/opentelemetry/integrationtest/OtlpExporterIntegrationTest.java#L837

In any case, I realized that I need the otel collector to test the second case you suggest below.

> * pre-canned mutableSpan -> existing sender + new encoder -> otlp -> readback span sent
> * pre-canned mutableSpan -> existing sender + zipkin encoder -> zipkin endpoint on collector with zipkin enabled -> readback span sent

To understand the context, could you please tell me from what perspective the tests are needed for the second case since the otel encoder is not being used?

**Comment by codefromthecrypt** [comment]
so for an encoder, test we don't really need a mock server. we need something that can deserialize the format we wrote, the format valid for the /v1/traces endpoint. I think it is better to focus on encoder test being about deserializing those bytes you just wrote with a native decoder (e.g. some otel classes in java that deserlalize the message exactly how the /v1/traces endpoint in the collector would). Adding a remote transport doesn't really help here.

The IT tests would be the only things that need a remote transport (e.g. hook the encoder to an existing http sender and point it at a testcontainers collector with readback as discussed earlier).

Does this make sense?

**Comment by codefromthecrypt** [comment]
So, your OtlpEncoder, I guess there are two variants json and proto. You can start with one of those, and then use the corresponding unmarshal function in your OtlpEncoderTest to read back what was written and verify the fields.

here's where the server is accepting content types which can be json or proto.

https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/otlphttp.go#L120-L135

depending on how you write the classes, the code in zipkin-gcp may be handy for examples of unit tests. Integration tests as I described earlier aren't in zipkin-gcp as it requires a login

https://github.com/openzipkin/zipkin-gcp/tree/master/encoder-stackdriver-brave/src/test/java/zipkin2/reporter/stackdriver/brave

**Comment by anuraaga** [comment]
Just a reference since hadn't noticed JSON here yet - AFAIK OTel java SDK among many others don't support it, I think the protocol only exists to support browsers and even that I don't know if it's used in practice.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#exporters

So probably don't need to support it in Zipkin unless it's trivial or something.

**Comment by making** [comment]
@codefromthecrypt 
> My suggestion is to work backwards from the integration test and make the least possible code to equate the following, or if not explain why.

Based on this comment, I'm writing the test like the following, but is this approach appropriate?

https://github.com/making/zipkin-otel/blob/sender/encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITOtelEncoderTest.java

It would be helpful if you could take a look at it when you have time.

**Comment by codefromthecrypt** [comment]
@making sorry I was out sick for several days and only now burned notifications down far enough to see your comment. Your approach is exactly what I was looking for. I might be more strict about the attribute assertions (like list equals vs one-by-one) so that we know if there is some accidental extra value etc. This is a minor point though.

File: encoder-otel-brave/pom.xml (hunk 2 above)
Comment (making) [reply encoder-otel-brave/pom.xml:1]: Fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/9f0dcdc630b2f8f1cbf422655a2fdcb54600e458

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/JaegerAllInOne.java (hunk 3 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/JaegerAllInOne.java:21]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

File: encoder-otel-brave/pom.xml (hunk 4 above)
Comment (making) [reply encoder-otel-brave/pom.xml:73]: Fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/ce421f0e29cea6c25d92167c61f96b378b67618b

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/AttributesExtractor.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import io.opentelemetry.proto.common.v1.AnyValue;
+import io.opentelemetry.proto.common.v1.KeyValue;
+import io.opentelemetry.proto.trace.v1.Span;
+import io.opentelemetry.proto.trace.v1.Status;
+import io.opentelemetry.proto.trace.v1.Status.StatusCode;
+import java.util.Map;
+
+/**
+ * LabelExtractor extracts the set of OTel Span labels equivalent to the annotations in a
+ * given Zipkin Span.
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/AttributesExtractor.java:27]: This class is inlined in `SpanTranslator`

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/AttributesExtractor.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+  AttributesExtractor(Tag<Throwable> errorTag, Map<String, String> renamedLabels) {
+    this.errorTag = errorTag;
+    this.renamedLabels = renamedLabels;
+  }
+
+  void addErrorTag(Span.Builder target, MutableSpan braveSpan) {
+    String errorValue = errorTag.value(braveSpan.error(), null);
+    if (errorValue != null) {
+      target.addAttributes(KeyValue.newBuilder().setKey(getLabelName("error")).setValue(
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/AttributesExtractor.java:51]: Fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/b42b81cc9bb96c264c68ccfe2f22d6f689b01e8f

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+  static final String OTEL_STATUS_CODE = "otel.status_code";
+
+  private static final Map<String, String> RENAMED_LABELS;
+
+  static {
+    RENAMED_LABELS = new LinkedHashMap<>();
+    RENAMED_LABELS.put("http.host", ServerAttributes.SERVER_ADDRESS.getKey());
+    RENAMED_LABELS.put("http.method", HttpAttributes.HTTP_REQUEST_METHOD.getKey());
+    RENAMED_LABELS.put("http.status_code", HttpAttributes.HTTP_RESPONSE_STATUS_CODE.getKey());
+    RENAMED_LABELS.put("http.request.size", "http.request.body.size");
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:69]: Fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/0fe808bf543dc30ce8d88a0da28aee6d57756f8a

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+
+/**
+ * SpanTranslator converts a Brave Span to a OpenTelemetry Span.
+ */
+final class SpanTranslator {
+
+  static final String KEY_INSTRUMENTATION_SCOPE_NAME = "otel.scope.name";
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:56]: `otel.scope.*` and `otel.library.*` attributed are removed as these are supposed to be used when converting from otel to non-otel. 
ImplementationScope is now used instead.

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+    if (peerIp != null) {
+      spanBuilder.addAttributes(KeyValue.newBuilder().setKey(SemanticAttributes.NET_SOCK_PEER_ADDR.getKey())
+          .setValue(AnyValue.newBuilder().setStringValue(peerIp).build()).build());
+    }
+    int peerPort = span.remotePort();
+    if (peerPort != 0) {
+      spanBuilder.addAttributes(KeyValue.newBuilder().setKey(SemanticAttributes.NET_SOCK_PEER_PORT.getKey())
+          .setValue(AnyValue.newBuilder().setIntValue(peerPort).build()).build());
+    }
+    // TODO: What should we put here?
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:176]: `net.sock.*` attributes are removed as these are deprecated.
`network.peer.*` are used instead.

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    @Override
+    public void accept(Span.Builder target, long timestamp, String value) {
+      target.addEvents(Event.newBuilder().setTimeUnixNano(TimeUnit.MICROSECONDS.toNanos(timestamp))
+          .setName(value).build());
+    }
+  }
+
+  static class LinkUtils {
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:227]: The link conversion is TBD, so deleted now.
https://opentelemetry.io/docs/specs/otel/trace/sdk_exporters/zipkin/

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+ */
+final class SpanTranslator {
+
+  static final String KEY_INSTRUMENTATION_SCOPE_NAME = "otel.scope.name";
+  static final String KEY_INSTRUMENTATION_SCOPE_VERSION = "otel.scope.version";
+  static final String KEY_INSTRUMENTATION_LIBRARY_NAME = "otel.library.name";
+  static final String KEY_INSTRUMENTATION_LIBRARY_VERSION = "otel.library.version";
+  static final String OTEL_STATUS_CODE = "otel.status_code";
+
+  private static final Map<String, String> RENAMED_LABELS;
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:62]: inlined in SpanTranslator

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+    RENAMED_LABELS.put("http.status_code", HttpAttributes.HTTP_RESPONSE_STATUS_CODE.getKey());
+    RENAMED_LABELS.put("http.request.size", "http.request.body.size");
+    RENAMED_LABELS.put("http.response.size", "http.response.body.size");
+    RENAMED_LABELS.put("http.url", UrlAttributes.URL_FULL.getKey());
+  }
+
+  private final Consumer consumer;
+
+  SpanTranslator(Tag<Throwable> errorTag) {
+    this.consumer = new Consumer(new AttributesExtractor(errorTag, RENAMED_LABELS));
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:77]: inlined in SpanTranslator

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+    String localServiceName = span.localServiceName();
+    if (localServiceName != null) {
+      resourceSpansBuilder.getResourceBuilder().addAttributes(
+          KeyValue.newBuilder().setKey(ServiceAttributes.SERVICE_NAME.getKey())
+              .setValue(AnyValue.newBuilder().setStringValue(localServiceName).build()).build());
+    } else {
+      resourceSpansBuilder.getResourceBuilder()
+          .addAttributes(KeyValue.newBuilder().setKey("service.name").setValue(
+              AnyValue.newBuilder().setStringValue(
+                  io.opentelemetry.sdk.resources.Resource.getDefault().getAttribute(stringKey("service.name"))).build()).build());
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:149]: Fixed

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+    }
+    String localIp = span.localIp();
+    if (localIp != null) {
+      spanBuilder.addAttributes(KeyValue.newBuilder().setKey(NetworkAttributes.NETWORK_LOCAL_ADDRESS.getKey())
+          .setValue(AnyValue.newBuilder().setStringValue(localIp).build()).build());
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:154]: fixed with `stringAttribute` method.
Sorry, I can't refer to a specific commit as there are several changes.

File: encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+    span.forEachAnnotation(consumer, spanBuilder);
+    consumer.addErrorTag(spanBuilder, span);
+
+    List<Link> links = LinkUtils.toLinks(span.tags());
+    spanBuilder.addAllLinks(links);
+
+    return spanBuilder;
+  }
+
+  class Consumer implements TagConsumer<Span.Builder>, AnnotationConsumer<Span.Builder> {
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/main/java/zipkin2/reporter/otel/brave/SpanTranslator.java:201]: fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/733b965a20de3ad7dde2a9bb0f84b77f6b127521

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/SpanTranslatorTest.java [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+                .setName("get")
+                .setKind(SpanKind.SPAN_KIND_CLIENT)
+                .setStartTimeUnixNano(
+                    TimeUnit.MILLISECONDS.toNanos(
+                        Instant.ofEpochSecond(1472470996, 199_000_000).toEpochMilli()))
+                .setEndTimeUnixNano(
+                    TimeUnit.MILLISECONDS.toNanos(
+                        Instant.ofEpochSecond(1472470996, 406_000_000).toEpochMilli()))
+                .addAllAttributes(
+                    Arrays.asList(KeyValue.newBuilder().setKey("network.local.address").setValue(
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/SpanTranslatorTest.java:64]: fixed in https://github.com/openzipkin-contrib/zipkin-otel/pull/7/commits/98748894f9cf6cd02b2f2267fbf3a9f6d5f163d0

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/TestObjects.java [hunk 23]
```diff
@@ -0,0 +1,41 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.reporter.otel.brave;
+
+import brave.handler.MutableSpan;
+
+public class TestObjects {
```
[the remark this answers is not in the document]
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/TestObjects.java:18]: Please let me confirm before deleting.
I believe this class was copied from zipkin-gcp.

https://github.com/openzipkin/zipkin-gcp/blob/master/encoder-stackdriver-brave/src/test/java/zipkin2/reporter/stackdriver/brave/TestObjects.java

Do you want me to inline it?

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 5 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:69]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 6 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:96]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 7 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:120]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 8 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:155]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 9 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:207]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3) and the comment has also been addressed

File: encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java (hunk 10 above)
Comment (making) [reply encoder-otel-brave/src/test/java/zipkin2/reporter/otel/brave/ITZipkinSenderTests.java:70]: This class is replaced by [ITOtelEncoderTest](https://github.com/openzipkin-contrib/zipkin-otel/pull/7/files#diff-f1614b313ae6e7ebe173556465333def6f893b421571522e9bb496a42ede83c3)

**Comment by making** [comment]
> I might be more strict about the attribute assertions (like list equals vs one-by-one) so that we know if there is some accidental extra value etc. 

Changed to directly compare lists with each other.

**Comment by making** [comment]
@codefromthecrypt 

Sorry for the late response (my main work became a bit busier).

I don't have permission to edit this branch directly, so I've responded to the comments in this PR and created a new PR at https://github.com/openzipkin-contrib/zipkin-otel/pull/7. I'd appreciate it if you could take a look and comment.

**Comment by reta** [comment]
I think we can close this one since https://github.com/openzipkin-contrib/zipkin-otel/pull/9 got merged, @marcingrzejszczak any objections from your side? Thank you!

**Comment by marcingrzejszczak** [comment]
Be my guest

**Pull request closed without merging**